command/flag.go

```diff
@@ -41,3 +44,123 @@ func (e EnumValue) String() string {
 func (e EnumValue) Get() interface{} {
 	return e
 }
+
+type MapValue struct {
+	m map[string]string
+}
+
+func NewMapValue() *MapValue {
+	return &MapValue{
+		m: map[string]string{},
+	}
+}
+
+func (m *MapValue) String() string {
+	s := ""
+
+	for key, value := range m.m {
+		s = fmt.Sprintf("%s, %s=%s, ", s, key, value)
+	}
+
+	return s
+}
+
+func (m *MapValue) Set(s string) error {
+	if len(s) == 0 {
+		return fmt.Errorf("flag can't be passed empty. Format: key=value")
+	}
+
+	tokens := strings.Split(s, "=")
+
+	if len(tokens) != 2 {
+		return fmt.Errorf("the key value pair(%s) format is invalid", tokens)
+	}
+
+	_, ok := m.m[tokens[0]]
+	if ok {
+		return fmt.Errorf("the key: %s is already defined", tokens[0])
+	}
+
+	m.m[tokens[0]] = tokens[1]
+
+	return nil
+}
+
+func (m *MapValue) Get() interface{} {
+	return m.m
+}
+
+type MapFlag struct {
+	Name string
+
+	Category    string
+	DefaultText string
+	FilePath    string
+	Usage       string
+
+	HasBeenSet bool
+	Required   bool
+	Hidden     bool
+
+	M *MapValue
+}
+
+func (f *MapFlag) Apply(set *flag.FlagSet) error {
+	f.M = NewMapValue()
+	for _, name := range f.Names() {
+		set.Var(f.M, name, f.Usage)
+		if len(f.M.m) > 0 {
+			f.HasBeenSet = true
+		}
+	}
+
+	return nil
+}
+
+func (f *MapFlag) GetUsage() string {
+	return f.Usage
+}
+
+func (f *MapFlag) Names() []string {
+	return []string{f.Name}
+}
+
+func (f *MapFlag) IsSet() bool {
+	return f.HasBeenSet
+}
+
+func (f *MapFlag) IsVisible() bool {
+	return true
+}
+
+func (f *MapFlag) String() string {
+	return cli.FlagStringer(f)
+}
+
+func (f *MapFlag) TakesValue() bool {
+	return true
+}
+
+func (f *MapFlag) GetValue() string {
+	return f.M.String()
+}
+
+func (f *MapFlag) GetDefaultText() string {
+	return ""
+}
+
+func (f *MapFlag) GetEnvVars() []string {
+	return []string{}
+}
+
+func (f *MapFlag) IsRequired() bool {
+	return f.Required
+}
+
+func (f *MapFlag) Get(cCtx *cli.Context) map[string]*string {
+	val, ok := cCtx.Generic(f.Name).(map[string]*string)
+	if !ok {
+		return nil
+	}
+	return val
+}
```

Comment:
Seems like removing this method does not break anything and causes no change on help output AFAIS, what it was required?

Reply:
What do you thing adding something like ;
```go
var (
	_ cli.Flag              = (*MapFlag)(nil)
	_ cli.RequiredFlag      = (*MapFlag)(nil)
	_ cli.VisibleFlag       = (*MapFlag)(nil)
	_ cli.DocGenerationFlag = (*MapFlag)(nil)
)
```

to assert that `MapFlat` implements them since we required to do for for things like required/help output etc. ?

It is kinda verbose but since all except `cli.Flag` is optional it is kinda hard to discover.

Reply:
The `Get` method is unused, I have mistakenly implemented it thinking that it's required for the `DocGenerationFlag` interface.  The thing you have proposed to put is very useful, I have added it too. Thanks for the new cool trick.  [a97c889](https://github.com/peak/s5cmd/pull/621/commits/a97c889e2ccd3cc09d11744d702bc9b5a3b5d4bf)

Reply:
Reference; https://github.com/uber-go/guide/blob/master/style.md#verify-interface-compliance